pkg/appliance/backup/backup.go

```diff
@@ -86,6 +90,73 @@ func (b *Backup) Download(ctx context.Context, applianceID, backupID, destinatio
 	return out, nil
 }
 
+// Download a completed appliance backup with the given ID of an Appliance
+func (b *Backup) Download(ctx context.Context, applianceID, backupID, destination string) (*os.File, error) {
+	out, err := os.Create(destination)
+	if err != nil {
+		return nil, err
+	}
+	defer out.Close()
+
+	client := b.HTTPClient
+	cfg := b.APIClient.GetConfig()
+	url, err := cfg.ServerURLWithContext(ctx, "ApplianceBackupApiService.AppliancesIdBackupBackupIdGet")
+	if err != nil {
+		return nil, err
+	}
+	url = fmt.Sprintf("%s/appliances/%s/backup/%s", url, applianceID, backupID)
+	ctx = context.WithValue(ctx, api.ContextAcceptValue, fmt.Sprintf("application/vnd.appgate.peer-v%d+gpg", b.Version))
+
+	headReq, err := http.NewRequestWithContext(ctx, http.MethodHead, url, http.NoBody)
+	if err != nil {
+		return nil, err
+	}
+	head, err := client.Do(headReq)
+	if err != nil {
+		return nil, err
+	}
+
+	size := head.ContentLength
+	if err := retryDownload(ctx, client, out, url, size); err != nil {
+		return nil, err
+	}
+
+	return out, nil
+}
+
+func retryDownload(ctx context.Context, client *http.Client, f *os.File, url string, size int64) error {
+	start := int64(0)
+	return backoff.Retry(func() error {
+		req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, http.NoBody)
+		if err != nil {
+			return backoff.Permanent(err)
+		}
+		req.Header.Set("Range", fmt.Sprintf("bytes=%d-", start))
+
+		log := logrus.WithField("request", req)
+		res, err := client.Do(req)
+		if err != nil {
+			return err
+		}
+		log.WithField("response", res).Debug("download request and response")
+		defer res.Body.Close()
+		if res.StatusCode >= 400 {
+			return fmt.Errorf("response does not indicate success: %v", res.Status)
+		}
+		buf := &bytes.Buffer{}
+		n, nerr := io.Copy(buf, res.Body)
+		_, err = f.WriteAt(buf.Bytes(), start)
```

Comment:
Is it possible to copy directly from the response to the file and skip the buffer altogether.
I assume the write call will return the number of bytes written, which is enough to advance start.

Reply:
The write call has a byte slice as a parameter, so it still needs to be copied from the body. And yes, it returns the number of bytes written, but we keep track of that in the copy call already and determine where it should start writing with that.

Reply:
Hmm ok i get the impression that will copy the entire body at once instead of doing it in chunks.
So a 10GB backup file would be fully be copied to the buffer in memory assuming no disconnects.

Reply:
Ping about my last comment here.

Reply:
It will keep a buffer in memory until an error occurs. If that happens, it will still write to the file and retry with a new buffer. Supposedly the old buffer will be garbage collected by then, freeing up the memory.

This is necessary because the API still does not support doing multiple concurrent ranged requests, so this will not download the backup in chunks, but it will try to resume the download in case of a failure. Unfortunately this makes it necessary to keep a big buffer in memory. At least until we're able to make concurrent requests. Natan is aware of this, so a fix for that will come in an API update later.

Reply:
I am not talking about doing chunks at the API level but rather reading in chunks from the socket level.
E.g. reading 1MB at a time (or the appropriate chunk size) from the socket and writing one chunk at a time to the file.
This way you only need a 1MB buffer regardless of the size of the backup, and regardless of how the API works.

If I understand how io.Copy works this would mean calling it directly from res.Body to f, something like `io.Copy(res.Body, f)` instead of using an intermediate buffer. And use the return value to increment the offset, see https://pkg.go.dev/io#Copy

See also https://stackoverflow.com/questions/11692860/how-can-i-efficiently-download-a-large-file-using-go

Reply:
I think that is the defacto way to avoid the excessive allocation of memory. The simple way is to use io.ReadAll() but the main drawback is that it reads the full input into memory, which may cause memory exhaustion and crashes if the input is too big.

https://betterprogramming.pub/the-hidden-risks-of-a-popular-way-to-read-http-requests-in-go-208a47481f8b
>The io.Copy function can be used as an alternative to io.ReadAll to copy data from an input source to an output source more efficiently by using less memory. This is a case to consider when we need to read files from an HTTP request. Because io.Copy reads data in small chunks, it performs better and lowers the possibility of running out of memory. It uses a 32-Kilobyte buffer to store data and flushes the contents of the buffer to disk.

Reply:
Yes, but `Copy` will always start to write from byte 0, meaning it will overwrite the bytes already written to file if it is called more than once and the backup will be corrupted for sure.

The stackoverflow thread concerns downloading large files efficiently, but not handling if the download fails midway. It just assumes it will, but midway download failures is the whole reason we're fixing this issue now. That's the reason why we need to use `WriteAt` in the first place, and that (and all other offset writers in go) needs a buffer.

Reply:
What can and should be be fixed is to stream the write though! Check the example in the link above